File: paykit-mobile/swift/KeyManager.swift
// KeyManager.swift
// Paykit iOS Key Management
//
// This file provides secure key management using the Rust FFI functions
// and iOS Keychain for persistent storage.
//
// Key Architecture:
//   - Ed25519 identity key (pkarr): The user's main identity
//   - X25519 device keys (pubky-noise): Derived for encryption
//
// USAGE:
//   let keyManager = KeyManager()
//   let keypair = try await keyManager.getOrCreateIdentity()
//   print("Your public key: \(keypair.publicKeyZ32)")

import Foundation

/// Manages cryptographic keys for Paykit
///
/// This class handles:
/// - Generating Ed25519 identity keys via Rust FFI
/// - Deriving X25519 device keys for Noise protocol
/// - Secure storage in iOS Keychain
/// - Key backup/restore with password encryption
public final class KeyManager: ObservableObject {
    
    // MARK: - Constants
    
    private enum Keys {
        static let secretKey = "paykit.identity.secret"
        static let publicKey = "paykit.identity.public"
        static let publicKeyZ32 = "paykit.identity.public.z32"
        static let deviceId = "paykit.device.id"
        static let currentEpoch = "paykit.device.epoch"
    }
    
    // MARK: - Published Properties
    
    @Published public private(set) var hasIdentity: Bool = false
    @Published public private(set) var publicKeyZ32: String = ""
    @Published public private(set) var publicKeyHex: String = ""
    
    // MARK: - Private Properties
    
    private let keychain: KeychainStorage
    private let deviceId: String
    
    // MARK: - Initialization
    
    public init(serviceIdentifier: String = Bundle.main.bundleIdentifier ?? "com.paykit.demo") {
        self.keychain = KeychainStorage(serviceIdentifier: serviceIdentifier)
        
        // Get or generate device ID
        if let storedDeviceId = try? keychain.retrieveString(key: Keys.deviceId) {
            self.deviceId = storedDeviceId
        } else {
            let newDeviceId = generateDeviceId()
            try? keychain.store(key: Keys.deviceId, string: newDeviceId)
            self.deviceId = newDeviceId
        }
        
        // Check for existing identity
        loadIdentityState()
    }
    
    // MARK: - Identity Management
    
    /// Get the current identity or create a new one
    ///
    /// - Returns: The Ed25519 keypair
    public func getOrCreateIdentity() throws -> Ed25519Keypair {
        if let existingSecret = try keychain.retrieveString(key: Keys.secretKey) {
            // Restore from storage
            return try ed25519KeypairFromSecret(secretKeyHex: existingSecret)
        } else {
            // Generate new identity
            return try generateNewIdentity()
        }
    }
    
    /// Generate a new identity (replaces existing if any)
    ///
    /// - Returns: The new Ed25519 keypair
    @discardableResult
    public func generateNewIdentity() throws -> Ed25519Keypair {
        let keypair = try generateEd25519Keypair()
        
        // Store in keychain
        try keychain.store(key: Keys.secretKey, string: keypair.secretKeyHex)
        try keychain.store(key: Keys.publicKey, string: keypair.publicKeyHex)
        try keychain.store(key: Keys.publicKeyZ32, string: keypair.publicKeyZ32)
        
        // Update published state
        DispatchQueue.main.async {
            self.hasIdentity = true
            self.publicKeyZ32 = keypair.publicKeyZ32
            self.publicKeyHex = keypair.publicKeyHex
        }
        
        return keypair
    }
    
    /// Get the current public key in z-base32 format (pkarr format)
    ///
    /// - Returns: The public key or nil if no identity exists
    public func getCurrentPublicKeyZ32() -> String? {
        try? keychain.retrieveString(key: Keys.publicKeyZ32)
    }
    
    /// Get the current public key in hex format
    ///
    /// - Returns: The public key or nil if no identity exists
    public func getCurrentPublicKeyHex() -> String? {
        try? keychain.retrieveString(key: Keys.publicKeyHex)
    }
    
    /// Delete the current identity
    ///
    /// Warning: This cannot be undone unless you have a backup!
    public func deleteIdentity() throws {
        try keychain.delete(key: Keys.secretKey)
        try keychain.delete(key: Keys.publicKey)
        try keychain.delete(key: Keys.publicKeyZ32)
        
        DispatchQueue.main.async {
            self.hasIdentity = false
            self.publicKeyZ32 = ""
            self.publicKeyHex = ""
        }
    }
    
    // MARK: - X25519 Device Keys
    
    /// Get or derive X25519 device key for Noise protocol
    ///
    /// - Parameter epoch: Key rotation epoch (default 0)
    /// - Returns: The X25519 keypair for this device and epoch
    public func getDeviceX25519Key(epoch: UInt32 = 0) throws -> X25519Keypair {
        guard let secretHex = try keychain.retrieveString(key: Keys.secretKey) else {
            throw KeyManagerError.noIdentity
        }
        
        return try deriveX25519Keypair(
            ed25519SecretHex: secretHex,
            deviceId: deviceId,
            epoch: epoch
        )
    }
    
    /// Get the current device ID
    public func getDeviceId() -> String {
        deviceId
    }
    
    /// Get the current key epoch
    public func getCurrentEpoch() -> UInt32 {
        guard let epochStr = try? keychain.retrieveString(key: Keys.currentEpoch),
              let epoch = UInt32(epochStr) else {
            return 0
        }
        return epoch
    }
    
    /// Increment the key epoch (for key rotation)
    public func incrementEpoch() throws -> UInt32 {
        let newEpoch = getCurrentEpoch() + 1
        try keychain.store(key: Keys.currentEpoch, string: String(newEpoch))
        return newEpoch
    }
    
    // MARK: - Signing
    
    /// Sign data with the identity key
    ///
    /// - Parameter data: Data to sign
    /// - Returns: Hex-encoded signature
    public func sign(data: Data) throws -> String {
        guard let secretHex = try keychain.retrieveString(key: Keys.secretKey) else {
            throw KeyManagerError.noIdentity
        }
        
        return try signMessage(secretKeyHex: secretHex, message: data)
    }
    
    /// Verify a signature
    ///
    /// - Parameters:
    ///   - publicKeyHex: Signer's public key in hex
    ///   - data: Original data
    ///   - signatureHex: Signature to verify
    /// - Returns: true if valid
    public func verify(publicKeyHex: String, data: Data, signatureHex: String) throws -> Bool {
        return try verifySignature(
            publicKeyHex: publicKeyHex,
            message: data,
            signatureHex: signatureHex
        )
    }
    
    // MARK: - Backup & Restore
    
    /// Export identity to encrypted backup
    ///
    /// - Parameter password: Password to encrypt the backup
    /// - Returns: The encrypted backup
    public func exportBackup(password: String) throws -> KeyBackup {
        guard let secretHex = try keychain.retrieveString(key: Keys.secretKey) else {
            throw KeyManagerError.noIdentity
        }
        
        return try exportKeypairToBackup(secretKeyHex: secretHex, password: password)
    }
    
    /// Import identity from encrypted backup
    ///
    /// - Parameters:
    ///   - backup: The encrypted backup
    ///   - password: Password to decrypt
    /// - Returns: The restored keypair
    @discardableResult
    public func importBackup(_ backup: KeyBackup, password: String) throws -> Ed25519Keypair {
        let keypair = try importKeypairFromBackup(backup: backup, password: password)
        
        // Store in keychain
        try keychain.store(key: Keys.secretKey, string: keypair.secretKeyHex)
        try keychain.store(key: Keys.publicKey, string: keypair.publicKeyHex)
        try keychain.store(key: Keys.publicKeyZ32, string: keypair.publicKeyZ32)
        
        // Update published state
        DispatchQueue.main.async {
            self.hasIdentity = true
            self.publicKeyZ32 = keypair.publicKeyZ32
            self.publicKeyHex = keypair.publicKeyHex
        }
        
        return keypair
    }
    
    /// Convert backup to shareable string (JSON)
    public func backupToString(_ backup: KeyBackup) throws -> String {
        let dict: [String: Any] = [
            "version": backup.version,
            "encrypted_data": backup.encryptedDataHex,
            "salt": backup.saltHex,
            "nonce": backup.nonceHex,
            "public_key": backup.publicKeyZ32
        ]
        let data = try JSONSerialization.data(withJSONObject: dict, options: [.prettyPrinted, .sortedKeys])
        return String(data: data, encoding: .utf8) ?? ""
    }
    
    /// Parse backup from string (JSON)
    public func backupFromString(_ string: String) throws -> KeyBackup {
        guard let data = string.data(using: .utf8),
              let dict = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let version = dict["version"] as? UInt32,
              let encryptedData = dict["encrypted_data"] as? String,
              let salt = dict["salt"] as? String,
              let nonce = dict["nonce"] as? String,
              let publicKey = dict["public_key"] as? String else {
            throw KeyManagerError.invalidBackup
        }
        
        return KeyBackup(
            version: version,
            encryptedDataHex: encryptedData,
            saltHex: salt,
            nonceHex: nonce,
            publicKeyZ32: publicKey
        )
    }
    
    // MARK: - Private Helpers
    
    private func loadIdentityState() {
        if let pubZ32 = try? keychain.retrieveString(key: Keys.publicKeyZ32),
           let pubHex = try? keychain.retrieveString(key: Keys.publicKey) {
            self.hasIdentity = true
            self.publicKeyZ32 = pubZ32
            self.publicKeyHex = pubHex
        } else {
            self.hasIdentity = false
            self.publicKeyZ32 = ""
            self.publicKeyHex = ""
        }
    }
}

// MARK: - Errors

public enum KeyManagerError: LocalizedError {
    case noIdentity
    case invalidBackup
    case backupDecryptionFailed
    
    public var errorDescription: String? {
        switch self {
        case .noIdentity:
            return "No identity key exists. Generate or import one first."
        case .invalidBackup:
            return "Invalid backup format."
        case .backupDecryptionFailed:
            return "Failed to decrypt backup. Check your password."
        }
    }
}

File: paykit-mobile/kotlin/KeyManager.kt
// KeyManager.kt
// Paykit Android Key Management
//
// This file provides secure key management using the Rust FFI functions
// and Android EncryptedSharedPreferences for persistent storage.
//
// Key Architecture:
//   - Ed25519 identity key (pkarr): The user's main identity
//   - X25519 device keys (pubky-noise): Derived for encryption
//
// USAGE:
//   val keyManager = KeyManager(context)
//   val keypair = keyManager.getOrCreateIdentity()
//   Log.d("Paykit", "Your public key: ${keypair.publicKeyZ32}")

package com.paykit.mobile

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONObject

/**
 * Manages cryptographic keys for Paykit
 *
 * This class handles:
 * - Generating Ed25519 identity keys via Rust FFI
 * - Deriving X25519 device keys for Noise protocol
 * - Secure storage in Android EncryptedSharedPreferences
 * - Key backup/restore with password encryption
 */
class KeyManager(context: Context) {
    
    companion object {
        private const val PREFS_NAME = "paykit_keys"
        private const val KEY_SECRET = "paykit.identity.secret"
        private const val KEY_PUBLIC = "paykit.identity.public"
        private const val KEY_PUBLIC_Z32 = "paykit.identity.public.z32"
        private const val KEY_DEVICE_ID = "paykit.device.id"
        private const val KEY_EPOCH = "paykit.device.epoch"
    }
    
    // State
    private val _hasIdentity = MutableStateFlow(false)
    val hasIdentity: StateFlow<Boolean> = _hasIdentity.asStateFlow()
    
    private val _publicKeyZ32 = MutableStateFlow("")
    val publicKeyZ32: StateFlow<String> = _publicKeyZ32.asStateFlow()
    
    private val _publicKeyHex = MutableStateFlow("")
    val publicKeyHex: StateFlow<String> = _publicKeyHex.asStateFlow()
    
    // Encrypted storage
    private val prefs: SharedPreferences
    private val deviceId: String
    
    init {
        // Create master key for encryption
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        
        // Create encrypted shared preferences
        prefs = EncryptedSharedPreferences.create(
            context,
            PREFS_NAME,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
        
        // Get or generate device ID
        deviceId = prefs.getString(KEY_DEVICE_ID, null) ?: run {
            val newId = generateDeviceId()
            prefs.edit().putString(KEY_DEVICE_ID, newId).apply()
            newId
        }
        
        // Load identity state
        loadIdentityState()
    }
    
    // ============================================================================
    // Identity Management
    // ============================================================================
    
    /**
     * Get the current identity or create a new one
     *
     * @return The Ed25519 keypair
     */
    fun getOrCreateIdentity(): Ed25519Keypair {
        val existingSecret = prefs.getString(KEY_SECRET, null)
        return if (existingSecret != null) {
            // Restore from storage
            ed25519KeypairFromSecret(existingSecret)
        } else {
            // Generate new identity
            generateNewIdentity()
        }
    }
    
    /**
     * Generate a new identity (replaces existing if any)
     *
     * @return The new Ed25519 keypair
     */
    fun generateNewIdentity(): Ed25519Keypair {
        val keypair = generateEd25519Keypair()
        
        // Store in encrypted preferences
        prefs.edit()
            .putString(KEY_SECRET, keypair.secretKeyHex)
            .putString(KEY_PUBLIC, keypair.publicKeyHex)
            .putString(KEY_PUBLIC_Z32, keypair.publicKeyZ32)
            .apply()
        
        // Update state
        _hasIdentity.value = true
        _publicKeyZ32.value = keypair.publicKeyZ32
        _publicKeyHex.value = keypair.publicKeyHex
        
        return keypair
    }
    
    /**
     * Get the current public key in z-base32 format (pkarr format)
     */
    fun getCurrentPublicKeyZ32(): String? = prefs.getString(KEY_PUBLIC_Z32, null)
    
    /**
     * Get the current public key in hex format
     */
    fun getCurrentPublicKeyHex(): String? = prefs.getString(KEY_PUBLIC, null)
    
    /**
     * Delete the current identity
     *
     * Warning: This cannot be undone unless you have a backup!
     */
    fun deleteIdentity() {
        prefs.edit()
            .remove(KEY_SECRET)
            .remove(KEY_PUBLIC)
            .remove(KEY_PUBLIC_Z32)
            .apply()
        
        _hasIdentity.value = false
        _publicKeyZ32.value = ""
        _publicKeyHex.value = ""
    }
    
    // ============================================================================
    // X25519 Device Keys
    // ============================================================================
    
    /**
     * Get or derive X25519 device key for Noise protocol
     *
     * @param epoch Key rotation epoch (default 0)
     * @return The X25519 keypair for this device and epoch
     */
    fun getDeviceX25519Key(epoch: UInt = 0u): X25519Keypair {
        val secretHex = prefs.getString(KEY_SECRET, null)
            ?: throw KeyManagerException.NoIdentity()
        
        return deriveX25519Keypair(
            ed25519SecretHex = secretHex,
            deviceId = deviceId,
            epoch = epoch
        )
    }
    
    /**
     * Get the current device ID
     */
    fun getDeviceId(): String = deviceId
    
    /**
     * Get the current key epoch
     */
    fun getCurrentEpoch(): UInt = prefs.getInt(KEY_EPOCH, 0).toUInt()
    
    /**
     * Increment the key epoch (for key rotation)
     */
    fun incrementEpoch(): UInt {
        val newEpoch = getCurrentEpoch() + 1u
        prefs.edit().putInt(KEY_EPOCH, newEpoch.toInt()).apply()
        return newEpoch
    }
    
    // ============================================================================
    // Signing
    // ============================================================================
    
    /**
     * Sign data with the identity key
     *
     * @param data Data to sign
     * @return Hex-encoded signature
     */
    fun sign(data: ByteArray): String {
        val secretHex = prefs.getString(KEY_SECRET, null)
            ?: throw KeyManagerException.NoIdentity()
        
        return signMessage(secretHex, data)
    }
    
    /**
     * Verify a signature
     *
     * @param publicKeyHex Signer's public key in hex
     * @param data Original data
     * @param signatureHex Signature to verify
     * @return true if valid
     */
    fun verify(publicKeyHex: String, data: ByteArray, signatureHex: String): Boolean {
        return verifySignature(publicKeyHex, data, signatureHex)
    }
    
    // ============================================================================
    // Backup & Restore
    // ============================================================================
    
    /**
     * Export identity to encrypted backup
     *
     * @param password Password to encrypt the backup
     * @return The encrypted backup
     */
    fun exportBackup(password: String): KeyBackup {
        val secretHex = prefs.getString(KEY_SECRET, null)
            ?: throw KeyManagerException.NoIdentity()
        
        return exportKeypairToBackup(secretHex, password)
    }
    
    /**
     * Import identity from encrypted backup
     *
     * @param backup The encrypted backup
     * @param password Password to decrypt
     * @return The restored keypair
     */
    fun importBackup(backup: KeyBackup, password: String): Ed25519Keypair {
        val keypair = importKeypairFromBackup(backup, password)
        
        // Store in encrypted preferences
        prefs.edit()
            .putString(KEY_SECRET, keypair.secretKeyHex)
            .putString(KEY_PUBLIC, keypair.publicKeyHex)
            .putString(KEY_PUBLIC_Z32, keypair.publicKeyZ32)
            .apply()
        
        // Update state
        _hasIdentity.value = true
        _publicKeyZ32.value = keypair.publicKeyZ32
        _publicKeyHex.value = keypair.publicKeyHex
        
        return keypair
    }
    
    /**
     * Convert backup to shareable string (JSON)
     */
    fun backupToString(backup: KeyBackup): String {
        return JSONObject().apply {
            put("version", backup.version)
            put("encrypted_data", backup.encryptedDataHex)
            put("salt", backup.saltHex)
            put("nonce", backup.nonceHex)
            put("public_key", backup.publicKeyZ32)
        }.toString(2)
    }
    
    /**
     * Parse backup from string (JSON)
     */
    fun backupFromString(string: String): KeyBackup {
        val json = JSONObject(string)
        return KeyBackup(
            version = json.getInt("version").toUInt(),
            encryptedDataHex = json.getString("encrypted_data"),
            saltHex = json.getString("salt"),
            nonceHex = json.getString("nonce"),
            publicKeyZ32 = json.getString("public_key")
        )
    }
    
    // ============================================================================
    // Private Helpers
    // ============================================================================
    
    private fun loadIdentityState() {
        val pubZ32 = prefs.getString(KEY_PUBLIC_Z32, null)
        val pubHex = prefs.getString(KEY_PUBLIC, null)
        
        if (pubZ32 != null && pubHex != null) {
            _hasIdentity.value = true
            _publicKeyZ32.value = pubZ32
            _publicKeyHex.value = pubHex
        } else {
            _hasIdentity.value = false
            _publicKeyZ32.value = ""
            _publicKeyHex.value = ""
        }
    }
}

// ============================================================================
// Exceptions
// ============================================================================

sealed class KeyManagerException(message: String) : Exception(message) {
    class NoIdentity : KeyManagerException("No identity key exists. Generate or import one first.")
    class InvalidBackup : KeyManagerException("Invalid backup format.")
    class BackupDecryptionFailed : KeyManagerException("Failed to decrypt backup. Check your password.")
}

